modules/juce_audio_devices/native/juce_mac_CoreAudio.cpp

```diff
@@ -376,16 +376,46 @@ class CoreAudioInternal  : private Timer
 
     int getBitDepthFromDevice (AudioObjectPropertyScope scope) const
     {
-        AudioObjectPropertyAddress pa;
-        pa.mElement = kAudioObjectPropertyElementMaster;
-        pa.mSelector = kAudioStreamPropertyPhysicalFormat;
-        pa.mScope = scope;
+        auto const streamsProperty = AudioObjectPropertyAddress{
+            kAudioDevicePropertyStreams,
+            scope,
+            kAudioObjectPropertyElementMaster
+        };
 
-        AudioStreamBasicDescription asbd;
-        UInt32 size = sizeof (asbd);
+        UInt32 size = 0;
+        if (OK(AudioObjectGetPropertyDataSize(deviceID, &streamsProperty, 0, nullptr, &size)) && (size > 0))
+        {
+            std::vector<UInt32> streamIds(size / sizeof(UInt32));
 
-        if (OK (AudioObjectGetPropertyData (deviceID, &pa, 0, nullptr, &size, &asbd)))
-            return (int) asbd.mBitsPerChannel;
+            if (OK(AudioObjectGetPropertyData(
+                deviceID, &streamsProperty, 0, nullptr, &size, streamIds.data())) && (size > 0))
+            {
+                UInt32 const numStreams = size / sizeof(UInt32);
+                std::vector<int> bitDepths(numStreams);

```

Comment:
I think using std::transform here adds cognitive complexity and compilers can't optimize it well (see https://godbolt.org/z/a403fs for toy example). If we're using functional style here, I think std::reduce with std::execution::seq (or homebrew implementation if we can't use C++17) would be a better fit.

Reply:
I don't want to overcomplicate things here, as this is not a critical code path...

Reply:
The thing is, in my opinion this solution is overcomplicated. As far as I understand what was attempted here was something like
```return max((bitsPerChannel(streamId)) for streamId in streamIds))```, but first we get the long lambda body because AudioObjectGetPropertyData has output parameters, then there's the extra busywork from C++ iterator interface, then there's the mutable array that should really not be needed (and you need to check if it's correct size) if this does what it looks like so you have to double check everything, and at this point there's nothing simple left. The compiler example is just pre-emptive counter to the standard argument that it doesn't matter because the compiler optimizes it anyway.

Reply:
And, well, there should be the check that the return value is valid. For bit depths 0 makes sense as non-value but it's something you have to think about when reading the code.